libgnucash/engine/Account.cpp

```diff
@@ -3178,54 +3130,85 @@ gnc_account_foreach_child (const Account *acc,
     }
 }
 
-void
-gnc_account_foreach_descendant (const Account *acc,
-                                AccountCb thunk,
-                                gpointer user_data)
+
+static gpointer
+account_foreach_descendant (const Account *acc, AccountCb2 thunk,
+                            gpointer user_data, gboolean sort, gboolean can_abort)
 {
     const AccountPrivate *priv;
-    GList *node;
-    Account *child;
+    gpointer result = nullptr;
+    GList *children;
 
-    g_return_if_fail(GNC_IS_ACCOUNT(acc));
-    g_return_if_fail(thunk);
+    g_return_val_if_fail (GNC_IS_ACCOUNT(acc), nullptr);
+    g_return_val_if_fail (thunk, nullptr);
 
     priv = GET_PRIVATE(acc);
-    for (node = priv->children; node; node = node->next)
+    if (sort)
     {
-        child = static_cast<Account*>(node->data);
-        thunk(child, user_data);
-        gnc_account_foreach_descendant(child, thunk, user_data);
+        children = g_list_copy (priv->children);
+        children = g_list_sort (children, (GCompareFunc)xaccAccountOrder);
     }
+    else
+        children = priv->children;
+
+    for (auto node = children; node; node = node->next)
+    {
+        auto child = static_cast<Account*>(node->data);
+        result = thunk (child, user_data);
+        if (can_abort && result) break;
+
+        result = account_foreach_descendant (child, thunk, user_data, sort, can_abort);
+        if (can_abort && result) break;
+    }
+
+    if (sort)
+        g_list_free (children);
+
+    return result;
 }
 
-gpointer
-gnc_account_foreach_descendant_until (const Account *acc,
-                                      AccountCb2 thunk,
-                                      gpointer user_data)
+typedef struct
```

Comment:
This is C++, no need for `typedef` and the name goes up top:
```
struct CollectAccounts
```

Reply:
> **Edit**: On the other hand there's no point in a one-element struct. Just pass the `GList*` directly.

That's the one thing I've been fighting and losing. How? There's `**` magic somewhere.

Reply:
Something like
```
GList* list = nullptr;
account_foreach_descendant (account, (AccountCb2)collect_acct, static_cast<gpointer>(&list), FALSE, FALSE);
```
 and 
```
auto listptr{static_cast<GList**>(user_data)};
*llistptr = g_list_prepend (*listptr, account);
```
Untested so it might take a little fiddling.

Reply:
Works😊 now `CountAccounts` could be eliminated too...

Reply:
\o/